### PR TITLE
MNT: Remove h5py2 compatibility code

### PR DESCRIPTION
**Pull request by kmuehlbauer**
- [x] Closes Issue #191
- [x] Tests fixed
- [x] Changes are documented in `CHANGELOG.rst` (needs finalizing in `README.rst`)

### CODE REVIEW

**Comment by kmuehlbauer** [comment]
@hmaarrfk It makes much sense to use NEP29 as you suggested. I'll add a note to README.rst after sorting out the CI issues here.

**Comment by hmaarrfk** [comment]
Not sure how far you want to with this. but conda-forge recently dropped python 3.7 and "nep29" would say to do the same.

**Comment by kmuehlbauer** [comment]
@hmaarrfk OK, so we could even drop Python 3.7. So, let's go for it.

**Comment by hmaarrfk** [comment]
I know your on it, but this also needs to get updated:
https://github.com/h5netcdf/h5netcdf/blob/main/setup.py#L36

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -74,7 +44,7 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        python-version: ["3.7", "3.8", "3.9", "3.10"]
+        python-version: ["3.8", "3.9", "3.10", "3.11"]
         h5py-version: ["3.6", "3.7"]
```
Comment (hmaarrfk) [inline .github/workflows/main.yml:48]:
```suggestion
        python-version: ["3.8", "3.9", "3.10"]
        h5py-version: ["3.6", "3.7"]
        include:
          - python-version: 3.11
            h5py-version: 3.7
```
https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#expanding-or-adding-matrix-configurations

**Comment by kmuehlbauer** [comment]
Thanks @hmaarrfk, I've applied the suggestion!

**Comment by kmuehlbauer** [comment]
@hmaarrfk I think everything is in place now, dropping h5py2 and python3.7 as well as mentioning of NEP29 and fixing CI. I'll merge by tomorrow latest, if there is nothing more to do.

**Comment by kmuehlbauer** [comment]
> Sounds great! thanks for this!

Then let's get this in, so you can rebase in your performance PR's.

**Pull request merged**